### PR TITLE
use structures to organize code

### PR DESCRIPTION
**Pull request by bzm3r**

### CODE REVIEW

File: src/15_hello_triangle.rs [hunk 1]
```diff
@@ -10,766 +10,730 @@ extern crate glsl_to_spirv;
...
-        for semaphore in semaphores_per_frame.into_iter() {
-            device.destroy_semaphore(semaphore);
-        }
+    fn create_command_pool(
+        device: &<back::Backend as Backend>::Device,
+        queue_type: queue::QueueType,
+        qf_id: queue::family::QueueFamilyId,
+    ) -> pool::CommandPool<back::Backend, Graphics> {
+        // raw command pool: a thin wrapper around command pools
+        // strongly typed command pool: a safe wrapper around command pools, which ensures that only one command buffer is recorded at the same time from the current queue
```
Comment (grovesNL) [inline src/15_hello_triangle.rs]: Let's split this comment into multiple lines

File: src/15_hello_triangle.rs [hunk 2]
```diff
@@ -10,766 +10,730 @@ extern crate glsl_to_spirv;
...
-            )];
-
-            let mut render_pass_inline_encoder = command_buffer.begin_render_pass_inline(
-                render_pass,
-                fb,
-                render_area,
-                clear_values.iter(),
-            );
-            render_pass_inline_encoder.draw(0..3, 0..1);
+struct HALState {
```
Comment (grovesNL) [inline src/15_hello_triangle.rs]: nit: `HalState` seems to be the convention

File: src/15_hello_triangle.rs [hunk 3]
```diff
@@ -10,766 +10,730 @@ extern crate glsl_to_spirv;
...
-}
 
-#[derive(Default)]
-struct QueueFamilyIds {
-    graphics_family: Option<hal::queue::QueueFamilyId>,
-}
+    fn create_framebuffers(
+        device: &<back::Backend as Backend>::Device,
+        render_pass: &<back::Backend as Backend>::RenderPass,
+        frame_images: &Vec<(
```
Comment (grovesNL) [inline src/15_hello_triangle.rs]: Anywhere we accept immutable `&Vec<>` we should accept a slice instead with `&[]` (https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#ptr_arg)

File: src/15_hello_triangle.rs [hunk 4]
```diff
@@ -10,766 +10,730 @@ extern crate glsl_to_spirv;
...
+        let format = formats.map_or(format::Format::Rgba8Srgb, |formats| {
+            formats
+                .iter()
+                .find(|format| format.base_format().1 == format::ChannelType::Srgb)
+                .map(|format| *format)
+                .unwrap_or(formats[0])
+        });
+
+        let swap_config = SwapchainConfig::from_caps(&caps, format);
+        let extent = swap_config.extent.clone();
```
Comment (grovesNL) [inline src/15_hello_triangle.rs]: This `clone` shouldn't be needed

File: src/15_hello_triangle.rs [hunk 5]
```diff
@@ -10,766 +10,730 @@ extern crate glsl_to_spirv;
...
+    render_finished_semaphores: Vec<<back::Backend as Backend>::Semaphore>,
+    image_available_semaphores: Vec<<back::Backend as Backend>::Semaphore>,
+    submission_command_buffers: Vec<command::Submit<back::Backend, Graphics, command::MultiShot, command::Primary>>,
+    command_pool: pool::CommandPool<back::Backend, Graphics>,
+    swapchain_framebuffers: Vec<<back::Backend as Backend>::Framebuffer>,
+    gfx_pipeline: <back::Backend as Backend>::GraphicsPipeline,
+    descriptor_set_layouts: Vec<<back::Backend as Backend>::DescriptorSetLayout>,
+    pipeline_layout: <back::Backend as Backend>::PipelineLayout,
+    render_pass: <back::Backend as Backend>::RenderPass,
+    frame_images: Vec<(<back::Backend as Backend>::Image, <back::Backend as Backend>::ImageView, )>,
```
Comment (grovesNL) [inline src/15_hello_triangle.rs]: Is the extra `, ` intentional?

**Review by grovesNL — COMMENTED** [review]
Awesome work! Looks great 👍

Also, if it makes it easier to use type aliases for repeated types like `GfxSwapchain` etc. as mentioned before, I think it's fine for now (hopefully we can get something upstream to replace these eventually).

File: src/15_hello_triangle.rs (hunk 1 above)
Comment (bzm3r) [reply src/15_hello_triangle.rs]: Done.

File: src/15_hello_triangle.rs (hunk 2 above)
Comment (bzm3r) [reply src/15_hello_triangle.rs]: Changed globally.

File: src/15_hello_triangle.rs (hunk 3 above)
Comment (bzm3r) [reply src/15_hello_triangle.rs]: Done.

File: src/15_hello_triangle.rs (hunk 4 above)
Comment (bzm3r) [reply src/15_hello_triangle.rs]: Fixed.

File: src/15_hello_triangle.rs [hunk 6]
```diff
@@ -10,766 +10,739 @@ extern crate glsl_to_spirv;
...
+            command_buffer.bind_graphics_pipeline(pipeline);
+            {
+                // begin render pass
+                let render_area = pso::Rect {
+                    x: 0,
+                    y: 0,
+                    w: extent.width as _,
+                    h: extent.height as _,
+                };
+                let clear_values = vec![command::ClearValue::Color(command::ClearColor::Float([
```
Comment (grovesNL) [inline src/15_hello_triangle.rs:591]: Not related to your change, but is this `4.0` supposed to be `1.0`?

File: src/15_hello_triangle.rs [hunk 7]
```diff
@@ -10,766 +10,739 @@ extern crate glsl_to_spirv;
...
 
-    device.destroy_command_pool(command_pool.into_raw());
+    fn create_command_buffers<'a>(
+        command_pool: &'a mut pool::CommandPool<back::Backend, Graphics>,
+        render_pass: &<back::Backend as Backend>::RenderPass,
+        framebuffers: &[<back::Backend as Backend>::Framebuffer],
+        extent: window::Extent2D,
+        pipeline: &<back::Backend as Backend>::GraphicsPipeline,
+    ) -> Vec<command::Submit<back::Backend, Graphics, command::MultiShot, command::Primary>> {
+        // reserve (allocate memory for) num number of primary command buffers
```
Comment (grovesNL) [inline src/15_hello_triangle.rs]: `num number`

File: src/15_hello_triangle.rs (hunk 6 above)
Comment (bzm3r) [reply src/15_hello_triangle.rs:591]: Yes, it is. I fixed it globally.

File: src/15_hello_triangle.rs (hunk 7 above)
Comment (bzm3r) [reply src/15_hello_triangle.rs]: Fixed.

**Review by grovesNL — APPROVED** [review]
Updates look great, thanks!
bors r+

**Pull request merged**